armory/art_experimental/attacks/dapricot_patch.py

```diff
@@ -1,23 +1,673 @@
+        ):
+            raise ValueError("The patch location must be outside the crop range.")
+
+        if not isinstance(self.brightness_range, (tuple, list)) or not all(
+            isinstance(s, float) for s in self.brightness_range
+        ):
+            raise ValueError(
+                "The brightness range must be either a tuple or list of floats."
+            )
+        if len(self.brightness_range) != 2:
+            raise ValueError("The length of brightness range must be 2.")
+
+        # if self.brightness_range[0] < 0.0 or self.brightness_range[1] > 1.0:
+        #     raise ValueError("The brightness range must be between 0.0 and 1.0.")
+
+        if self.brightness_range[0] > self.brightness_range[1]:
+            raise ValueError(
+                "The first element of the brightness range must be less or equal to the second one."
+            )
+
+        if not isinstance(self.rotation_weights, (tuple, list)) or not all(
+            isinstance(s, (float, int)) for s in self.rotation_weights
+        ):
+            raise ValueError(
+                "The rotation sampling weights must be provided as tuple or list of float or int values."
+            )
+        if len(self.rotation_weights) != 4:
+            raise ValueError("The number of rotation sampling weights must be 4.")
+
+        if not all(s >= 0.0 for s in self.rotation_weights):
+            raise ValueError("The rotation sampling weights must be non-negative.")
+
+        if all(s == 0.0 for s in self.rotation_weights):
+            raise ValueError(
+                "At least one of the rotation sampling weights must be strictly greater than zero."
+            )
+
+        if not isinstance(self.sample_size, int):
+            raise ValueError("The EOT sample size must be of type int.")
+        if self.sample_size <= 0:
+            raise ValueError("The EOT sample size must be greater than 0.")
+
+        if not isinstance(self.targeted, bool):
+            raise ValueError("The argument `targeted` has to be of type bool.")
+
+    def _augment_images_with_patch(self, x, y, patch, channels_first):
+        """
+        Augment images with patch using perspective transform
+        instead of inserting patch based on patch_location
+
+        :param x: Sample images.
+        :param y: Target labels.
+        :param patch: The patch to be applied.
+        :param channels_first: Set channels first or last.
+        """
+
+        if x.shape[0] != len(self.gs_coords):
+            raise ValueError(
+                "Number of images should be equal to the number of arrays of green screen coordinates"
+            )
+        if y is not None and (x.shape[0] != len(y)):
+            raise ValueError(
+                "Number of images should be equal to the number of targets"
+            )
+
+        transformations = dict()
+        x_copy = x.copy()
+        patch_copy = patch.copy()
+        x_patch = x.copy()
+
+        if channels_first:
```

Comment:
I think I may see an issue with this code if channels_first is True:

The value that gets passed for the `channels_first` arg is `self.estimator.channels_first`. Regardless of whether this is True, the `x` will always be channels_last when using Armory for dataset loading. So if `self.estimator.channels_first` is True, then we'll be calling `np.transpose(x_copy, (0, 2, 3, 1))` when `x_copy` is of shape `(3, 1008, 756, 3)` which will lead to an array of shape `(3, 756, 3, 1008)` which would be a problem

Reply:
I see. `self.estimator.channels_first == True` implies the model assumes the data has channel first, which isn't the case for the DAPRICOT dataset, or any dataset from Armory for that matter. The burden is on the model to transpose the data correctly and not on ART.

Reply:
correct